docs/specs/tsd-006-startup-error-behavior.md

```diff
@@ -0,0 +1,105 @@
+# Startup Error Behavior
+
+## Objective
+
+Unified, configurable behavior on retry-able startup errors.
+
+## Keywords
+
+inputs, outputs, startup, error, retry
+
+## Overview
+
+Many Telegraf plugins connect to an external service either on the same machine
+or via network. On automated startup of Telegraf (e.g. via service) there is no
+guarantee that those services are fully started yet, especially when they reside
+on a remote host. More and more plugins implement mechanisms to retry reaching
+their related service if they failed to do so on startup.
+
+This specification intends to unify the naming of configuration-options, the
+values of those options and their semantic meaning. It describes the behavior
+for the different options on handling startup-errors.
+
+Startup-errors are all errors occurring in calls to `Start()`, for inputs and
+service-inputs, or `Connect()`, for outputs. The behaviors described below
+should only be applied in cases where the plugin *explicitly* states that an
+startup-error is *retry-able*. This comprises for example network errors
+indicating that the host or service is not reachable (yet) or external
+resources, like a machine or file, which are not yet available but might become
+available later. To indicate a retry-able startup-error the plugin should return
+a predefined error-type, potentially including an explicit flag. Additionally,
+the plugin might indicate partial connects to signal
```

Comment:
I know we talked about this at the face-to-face, but how did we decide a plugin would decide if something was retriable? Each plugin would need to learn to parse the errors and make a determination or in the case of HTTP codes, retry if we could not connect, but not if it is the wrong password for example?

Reply:
Yeah exactly. Someone needs to decide if we should proceed or stop hammering a service. This _must_ be a per-plugin decision as we cannot know in the agent about every single possible error of a plugin (and the underlying library). By default, all non-specific errors are interpreted as not-retryable. In my memory we discussed this face-to-face, but my memory is weak. ;-)

Reply:
> my memory is weak. ;-)

You and me both :) hence why I had to ask this again. My hesitation is the current input plugins that have retry built-in, have assumed all errors are retriable. Switching to how we have this in the spec, could break them, no?

Reply:
No because their current approach is completely parallel to the framework. When converting them, we might either wrap all errors into a retry-able error _or_ differentiate dependent on the error cause.

Reply:
ok I'd like for us to talk through a real example then as far as how we would distinguish between retryable and non-retryable. For example, take the plugin kafka-consumer. What would a non-retryable error be? The only thing I can think of is if we specifically looked for invalid credentials? Otherwise, anything related to connecting would be retryable no?

So let's say a unauthorized error comes back what is the behavior with the different options:

* error - throw an error as we do now, no change
* retry - do we not retry and throw an error like we do now? Again as a user if I have gone through the effort of saying to retry, then I expect you to retry.

Reply:
ok after our discussion from yesterday, I think the only thing I'd like to see added here is:

* Specify that plugins could allow users to specify what is retryable or non-retryable
* What is the purpose of the explicit flag? I didn't recall this from yesterday
* Is the last sentence cut off?

Reply:
Cut-off sentence removed and documented ability for the plugin to define what a retryable error is.

For the flag, currently the error is defined as
```golang
type StartupError struct {
	Err   error
	Retry bool
}
```

so `Retry` is the explicit flag. Alternatively we could define

```golang
type RetryableStartupError struct {
	Err   error
}
```

or should I leave this out of the spec completely as it is a implementation detail?

Reply:
Let's leave this out. I think the flag may make sense, but this is too in the weeds for the specs.